File: recursion/javascript/solutions/foldl.js
/**
 * Solution for foldl.
 *
 * Left fold: processes left-to-right.
 *
 * foldl(f, z, [a, b, c]) = f(f(f(z, a), b), c)
 */

const { EMPTY_LIST, Node, isEmpty, unprepend } = require('./linkedList');

function foldl(fn, acc, list) {
  if (isEmpty(list)) {
    return acc;
  }

  let [first, rest] = unprepend(list);

  return foldl(fn, fn(acc, first), rest);
}

if (require.main === module) {
  // For associative operations, foldl and foldr give same result
  console.assert(foldl((acc, x) => acc + x, 0, Node.fromArray([1, 2, 3, 4, 5])) === 15);
  console.assert(foldl((acc, x) => acc * x, 1, Node.fromArray([1, 2, 3, 4, 5])) === 120);

  // For non-associative operations, order matters
  // foldl(-, 0, [1,2,3]) = ((0-1)-2)-3 = -6
  console.assert(foldl((acc, x) => acc - x, 0, Node.fromArray([1, 2, 3])) === -6);

  // Building a list: foldl reverses, foldr preserves
  // foldl with prepend: (([] . 1) . 2) . 3 = [3, 2, 1]
  const { prepend, toArray } = require('./linkedList');
  let reversed = foldl((acc, x) => prepend(x, acc), EMPTY_LIST, Node.fromArray([1, 2, 3]));
  console.assert(JSON.stringify(toArray(reversed)) === '[3,2,1]');

  console.log('All tests passed!');
}

module.exports = { foldl };

File: recursion/python/solutions/foldl.py
"""Solution for foldl."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def foldl(fn, acc, lst):
    """
    Left fold: iterative-style accumulator.

    foldl(fn, acc, a -> b -> c) = fn(fn(fn(acc, a), b), c)

    This is "iterative" in flavor - acc is updated as we go.
    Compare to foldr which is "recursive" in flavor.

    foldl(add, 0, 1 -> 2 -> 3) = ((0 + 1) + 2) + 3 = 6
    foldl(subtract, 0, 1 -> 2 -> 3) = ((0 - 1) - 2) - 3 = -6
    """
    if is_empty(lst):
        return acc

    first, rest = unprepend(lst)

    return foldl(fn, fn(acc, first), rest)


def add(a, b):
    return a + b


def subtract(a, b):
    return a - b


if __name__ == '__main__':
    # Test cases
    assert foldl(add, 0, Node.from_list([1, 2, 3])) == 6
    assert foldl(subtract, 0, Node.from_list([1, 2, 3])) == -6
    assert foldl(add, 0, EMPTY_LIST) == 0

    # Reverse via foldl
    assert to_python_list(foldl(lambda acc, x: prepend(x, acc), EMPTY_LIST, Node.from_list([1, 2, 3]))) == [3, 2, 1]

    # Sum with initial value
    assert foldl(add, 10, Node.from_list([1, 2, 3])) == 16

    print("All tests passed!")
